src/libraries/Microsoft.Extensions.Diagnostics/src/DefaultMeterFactory.cs

```diff
@@ -0,0 +1,126 @@
+                    {
+                        if (meter.Version == options.Version && CompareTags(meter.Tags, options.Tags))
+                        {
+                            return meter;
+                        }
+                    }
+                }
+                else
+                {
+                    meterList = new List<Meter>();
+                    _cachedMeters.Add(options.Name, meterList);
+                }
+
+                Meter m = new Meter(options.Name, options.Version, options.Tags, scope: this);
+                meterList.Add(m);
+                return m;
+            }
+        }
+
+        public void Dispose()
+        {
+            lock (_cachedMeters)
+            {
+                _disposed = true;
+
+                foreach (List<Meter> meterList in _cachedMeters.Values)
+                {
+                    foreach (Meter meter in meterList)
+                    {
+                        meter.Dispose();
+                    }
+                }
+
+                _cachedMeters.Clear();
+            }
+        }
+
+        private static bool CompareTags(IEnumerable<KeyValuePair<string, object?>>? tags1, IEnumerable<KeyValuePair<string, object?>>? tags2)
+        {
+            if (tags1 is null)
+            {
+                return tags2 is null;
+            }
+
+            if (tags2 is null)
+            {
+                return false;
+            }
+
+            if (tags1 is ICollection<KeyValuePair<string, object?>> firstCol && tags2 is ICollection<KeyValuePair<string, object?>> secondCol)
+            {
+                if (firstCol.Count != secondCol.Count)
+                {
+                    return false;
+                }
+
+                if (firstCol is IList<KeyValuePair<string, object?>> firstList && secondCol is IList<KeyValuePair<string, object?>> secondList)
+                {
+                    int count = firstList.Count;
+                    for (int i = 0; i < count; i++)
+                    {
+                        KeyValuePair<string, object?> pair1 = firstList[i];
+                        KeyValuePair<string, object?> pair2 = secondList[i];
+                        if (pair1.Key != pair2.Key || !object.Equals(pair1.Value, pair2.Value))
+                        {
+                            return false;
+                        }
+                    }
+
+                    return true;
+                }
+            }
+
+            using (IEnumerator<KeyValuePair<string, object?>> e1 = tags1.GetEnumerator())
+            using (IEnumerator<KeyValuePair<string, object?>> e2 = tags2.GetEnumerator())
+            {
+                while (e1.MoveNext())
+                {
+                    KeyValuePair<string, object?> pair1 = e1.Current;
+                    if (!e2.MoveNext() || pair1.Key != e2.Current.Key || !object.Equals(pair1.Value, e2.Current.Value))
+                    {
+                        return false;
+                    }
```

Comment:
```suggestion
                    if (!e2.MoveNext())
                    {
                        return false;
                    }
                    
                    KeyValuePair<string, object?> pair2 = e2.Current;
                    if (pair1.Key != pair2.Key || !object.Equals(pair1.Value, pair2.Value))
                    {
                        return false;
                    }
```
to avoid an extra interface call.